prometheus/cache.go

```diff
@@ -0,0 +1,236 @@
+package prometheus
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	kialiConfig "github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+)
+
+type (
+	timeInResult struct {
+		queryTime time.Time
+		inResult  model.Vector
+	}
+
+	timeInOutResult struct {
+		queryTime time.Time
+		inResult  model.Vector
+		outResult model.Vector
+	}
+
+	PromCache interface {
+		GetAllRequestRates(namespace string, ratesInterval string, queryTime time.Time) (bool, model.Vector)
+		GetAppRequestRates(namespace, app, ratesInterval string, queryTime time.Time) (bool, model.Vector, model.Vector)
+		GetNamespaceServicesRequestRates(namespace string, ratesInterval string, queryTime time.Time) (bool, model.Vector)
+		GetServiceRequestRates(namespace, service, ratesInterval string, queryTime time.Time) (bool, model.Vector)
+		GetWorkloadRequestRates(namespace, workload, ratesInterval string, queryTime time.Time) (bool, model.Vector, model.Vector)
+		SetAllRequestRates(namespace string, ratesInterval string, queryTime time.Time, inResult model.Vector)
+		SetAppRequestRates(namespace, app, ratesInterval string, queryTime time.Time, inResult model.Vector, outResult model.Vector)
+		SetNamespaceServicesRequestRates(namespace string, ratesInterval string, queryTime time.Time, inResult model.Vector)
+		SetServiceRequestRates(namespace, service, ratesInterval string, queryTime time.Time, inResult model.Vector)
+		SetWorkloadRequestRates(namespace, workload, ratesInterval string, queryTime time.Time, inResult model.Vector, outResult model.Vector)
+	}
+
+	promCacheImpl struct {
+		cacheDuration          time.Duration
+		cacheAllRequestRates   map[string]map[string]timeInResult
+		cacheAppRequestRates   map[string]map[string]map[string]timeInOutResult
+		cacheNsSvcRequestRates map[string]map[string]timeInResult
+		cacheSvcRequestRates   map[string]map[string]map[string]timeInResult
+		cacheWkRequestRates    map[string]map[string]map[string]timeInOutResult
```

Comment:
What is the advantage of map-of-maps[-of-maps] as opposed to constructing a single key and using a single map?

Reply:
Several advantages:
- It structures caching per categories (namespace, item, rate interval).
- It allows to invalidate / refresh per namespace (not yet introduced in this iteration but something to consider later).
- It saves to manipulate and build single primary keys for items.
- It follows similar design as used in kubernetes/caching.